core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -101,6 +101,53 @@ public static Snapshot oldestAncestor(Table table) {
     return ancestorsOf(start, lookup);
   }
 
+  /**
+   * Traverses the history of the table's current snapshot
+   * and finds the oldest ancestor snapshot after or equal to the timestamp in milliseconds.
+   *
+   * @param table           a table
+   * @param timestampMillis a timestamp in milliseconds
+   * @return the first snapshot from the given timestamp, or null if the current snapshot is older than the timestamp
+   * @throws IllegalStateException if the first snapshot from the given time can't be determined
+   */
+  public static Snapshot oldestAncestorFrom(Table table, Long timestampMillis) {
+    Snapshot currentSnapshot = table.currentSnapshot();
+    long timestamp = timestampMillis == null ? -1L : timestampMillis;
+    // If no snapshot exists or timestamp is higher than the current snapshot
+    if (currentSnapshot == null || currentSnapshot.timestampMillis() < timestamp) {
+      return null;
+    }
+
+    // Traverse to the oldest snapshot after or equal to the timestamp
+    Snapshot lastSnapshot = null;
+    for (Snapshot snapshot : currentAncestors(table)) {
+      if (snapshot.timestampMillis() == timestamp) {
+        return snapshot;
+      } else if (snapshot.timestampMillis() < timestamp) {
+        return lastSnapshot;
```

Comment:
So this is returning the the first snapshot created after the target timestamp

Reply:
Yes, it should be `timestamp >= targetTimestamp`

Reply:
Yes, you are correct. I have pushed a minimal version of this logic in the recent commit:

```
// Return the first snapshot which satisfies >= targetTimeStamp
Snapshot lastSnapshot = null;
for (Snapshot snapshot : currentAncestors(table)) {
  if (snapshot.timestampMillis() < targetTimeStamp) {
    return lastSnapshot;
  }
  lastSnapshot = snapshot;
}

// Return the oldest snapshot if the target timestamp is less than the oldest snapshot of the table
return lastSnapshot;
```